pkg/authorization/global_authorizer.go

```diff
@@ -33,33 +34,42 @@ import (
 )
 
 type GlobalAuthorizer struct {
-	newAuthorizer func(clusterName logicalcluster.Name) authorizer.Authorizer
+	globalRoleLister rbacv1listers.RoleClusterLister
+	localRoleLister  rbacv1listers.RoleClusterLister
+
+	globalRoleBindingLister rbacv1listers.RoleBindingClusterLister
+
+	globalClusterRoleLister rbacv1listers.ClusterRoleClusterLister
+	localClusterRoleLister  rbacv1listers.ClusterRoleClusterLister
+
+	globalClusterRoleBindingLister rbacv1listers.ClusterRoleBindingClusterLister
 }
 
 func NewGlobalAuthorizer(localKubeInformers, globalKubeInformers kcpkubernetesinformers.SharedInformerFactory) (authorizer.Authorizer, authorizer.RuleResolver) {
+	// listers are saved in the struct here to ensure that informers are started and we do not encounter race conditions with them.
 	a := &GlobalAuthorizer{
-		newAuthorizer: func(clusterName logicalcluster.Name) authorizer.Authorizer {
-			return rbac.New(
-				&rbac.RoleGetter{Lister: rbacwrapper.NewMergedRoleLister(
-					globalKubeInformers.Rbac().V1().Roles().Lister().Cluster(clusterName),
-					localKubeInformers.Rbac().V1().Roles().Lister().Cluster(controlplaneapiserver.LocalAdminCluster),
-				)},
-				&rbac.RoleBindingLister{Lister: globalKubeInformers.Rbac().V1().RoleBindings().Lister().Cluster(clusterName)},
-				&rbac.ClusterRoleGetter{Lister: rbacwrapper.NewMergedClusterRoleLister(
-					globalKubeInformers.Rbac().V1().ClusterRoles().Lister().Cluster(clusterName),
-					localKubeInformers.Rbac().V1().ClusterRoles().Lister().Cluster(controlplaneapiserver.LocalAdminCluster),
-				)},
-				&rbac.ClusterRoleBindingLister{Lister: globalKubeInformers.Rbac().V1().ClusterRoleBindings().Lister().Cluster(clusterName)},
-			)
-		},
+		globalRoleLister: globalKubeInformers.Rbac().V1().Roles().Lister(),
```

Comment:
nit: globalRoles

Would be less noisy.

Reply:
I tried to keep it consistent with `LocalAuthorizer`, which looks like this:

```go
a := &LocalAuthorizer{
    // listers are saved in the struct here to ensure that informers are started and we do not encounter race conditions with them.
    roleLister:               versionedInformers.Rbac().V1().Roles().Lister(),
    roleBindingLister:        versionedInformers.Rbac().V1().RoleBindings().Lister(),
    clusterRoleLister:        versionedInformers.Rbac().V1().ClusterRoles().Lister(),
    clusterRoleBindingLister: versionedInformers.Rbac().V1().ClusterRoleBindings().Lister(),
}
```